// xkeys/dist/__mocks__/node-hid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setDriverType = exports.devices = exports.HID = exports.setMockWriteHandler = void 0;
const events_1 = require("events");
let mockWriteHandler = undefined;
function setMockWriteHandler(handler) {
    mockWriteHandler = handler;
}
exports.setMockWriteHandler = setMockWriteHandler;
// export class HID extends EventEmitter {
class HID extends events_1.EventEmitter {
    constructor(_path) {
        super();
        this.mockWriteHandler = mockWriteHandler;
    }
    // constructor(vid: number, pid: number);
    close() {
        // void
    }
    pause() {
        // void
    }
    read(_callback) {
        // void
    }
    readSync() {
        return [];
    }
    readTimeout(_timeOut) {
        return [];
    }
    sendFeatureReport(_data) {
        return 0;
    }
    getFeatureReport(_reportIdd, _reportLength) {
        return [];
    }
    resume() {
        // void
    }
    write(message) {
        var _a;
        (_a = this.mockWriteHandler) === null || _a === void 0 ? void 0 : _a.call(this, this, message);
        return 0;
    }
    setNonBlocking(_noBlock) {
        // void
    }
}
exports.HID = HID;
function devices() {
    return [];
}
exports.devices = devices;
function setDriverType(_type) {
    // void
}
exports.setDriverType = setDriverType;
