// casparcg-connection/dist/lib/event/CasparCGSocketCommandEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const BaseEvent_1 = require("./BaseEvent");
var Event;
(function (Event) {
    /**
     *
     */
    class CasparCGSocketCommandEvent extends BaseEvent_1.Event.BaseEvent {
        /**
         *
         */
        constructor(command) {
            super({ command });
            this.command = command;
        }
        /**
         *
         */
        valueOf() {
            return this.command;
        }
    }
    CasparCGSocketCommandEvent.RESPONSE = 'casparcgsocketcommandeventresponse';
    Event.CasparCGSocketCommandEvent = CasparCGSocketCommandEvent;
    /**
     *
     */
    class CasparCGSocketResponseEvent extends BaseEvent_1.Event.BaseEvent {
        /**
         *
         */
        constructor(response) {
            super({ response });
            this.response = response;
        }
        /**
         *
         */
        valueOf() {
            return this.response;
        }
    }
    CasparCGSocketResponseEvent.RESPONSE = 'casparcgsocketresponseeventresponse';
    CasparCGSocketResponseEvent.INVALID_RESPONSE = 'casparcgsocketcommandeventresponseinvalid';
    Event.CasparCGSocketResponseEvent = CasparCGSocketResponseEvent;
})(Event = exports.Event || (exports.Event = {}));
