// casparcg-connection/dist/lib/AbstractCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// ResponseNS
const ResponseSignature_1 = require("./ResponseSignature");
var ResponseSignature = ResponseSignature_1.Response.ResponseSignature;
// Validation ND
const ParamValidators_1 = require("./ParamValidators");
var PositiveNumberValidatorBetween = ParamValidators_1.Validation.PositiveNumberRoundValidatorBetween;
/**
 *
 */
var Command;
(function (Command) {
    /**
     *
     */
    class AMCPResponse {
        toString() {
            return this.raw.replace(/\r?\n|\r/gi, '');
        }
    }
    Command.AMCPResponse = AMCPResponse;
    /**
     *
     */
    let IAMCPStatus;
    (function (IAMCPStatus) {
        IAMCPStatus[IAMCPStatus["Invalid"] = -1] = "Invalid";
        IAMCPStatus[IAMCPStatus["New"] = 0] = "New";
        IAMCPStatus[IAMCPStatus["Initialized"] = 1] = "Initialized";
        IAMCPStatus[IAMCPStatus["Queued"] = 2] = "Queued";
        IAMCPStatus[IAMCPStatus["Sent"] = 3] = "Sent";
        IAMCPStatus[IAMCPStatus["Suceeded"] = 4] = "Suceeded";
        IAMCPStatus[IAMCPStatus["Failed"] = 5] = "Failed";
        IAMCPStatus[IAMCPStatus["Timeout"] = 6] = "Timeout";
    })(IAMCPStatus = Command.IAMCPStatus || (Command.IAMCPStatus = {}));
    /**
     *
     */
    class AbstractCommand {
        // @todo: add concept of "variants", adding an ENUM to variants of the same (query) verb-command. INFO x INFO y, but not Thumbnail Retriece and thumbnail generate, different verbs
        // not LOG (action, not query)
        // INFO, HELP
        // @todo:
        // channel vs layer-specific vs layer-fallback addresses
        // NB.: INFO BOTH LAYER AND CHANNEL!!!!!!!!
        // INFO, SWAP, REMOVE, MIXER CLEAR, CLEAR,
        // param getter/setters
        // param list (dynamic)
        // media info/template file-type to generate param data for fields
        /**
         *
         */
        constructor(params, context) {
            this.context = context;
            this.response = new AMCPResponse();
            this.responseProtocol = new ResponseSignature();
            this._payload = {};
            this._status = IAMCPStatus.New;
            // parse params to objects
            let paramsArray = [];
            // conform params to array
            if (Array.isArray(params)) {
                paramsArray = params;
            }
            else {
                paramsArray = [params];
            }
            this._stringParamsArray = [];
            this._objectParams = {};
            this._token = Math.random().toString(35).substr(2, 7);
            for (let element of paramsArray) {
                if (typeof element === 'string') {
                    element = element.toString().trim();
                    this._stringParamsArray = this._stringParamsArray.concat([...element.toString().split(/\s+/)]); // @todo: string delimiter pairing (,;) -> objectArray
                }
                else {
                    for (let prop in element) {
                        this._objectParams[prop] = element[prop];
                    }
                }
            }
        }
        /**
         *
         */
        validateParams() {
            let required = this.paramProtocol ? this.paramProtocol.filter(signature => signature.required.valueOf() === true) : [];
            let optional = this.paramProtocol ? this.paramProtocol.filter(signature => signature.required.valueOf() === false) : [];
            // check all required
            for (let signature of required) {
                if (!this.validateParam(signature)) {
                    return false;
                }
            }
            // add valid optionals
            optional.forEach((signature) => {
                this.validateParam(signature);
            });
            if (!this.validateProtocolLogic()) {
                return false;
            }
            let validParams = this.paramProtocol ? this.paramProtocol.filter((param) => param.resolved && param.payload !== null) : [];
            let invalidParams = this.paramProtocol ? this.paramProtocol.filter((param) => param.resolved && param.payload === null && param.required.valueOf() === true) : [];
            if (invalidParams.length > 0) {
                return false;
            }
            validParams.forEach((param) => {
                let payload = { key: '', value: {}, raw: null };
                payload.key = param.key || '';
                payload.value = param.payload !== null ? param.payload : {};
                payload.raw = param.raw;
                this.payload[param.name] = payload;
            });
            return true;
        }
        getParam(name) {
            if (this._objectParams[name]) {
                return this._objectParams[name];
            }
            return undefined;
        }
        /**
         *
         */
        validateResponse(response) {
            // assign raw response
            this.response.raw = response.responseString;
            this.response.code = response.statusCode;
            // code is correct
            if (response.statusCode !== this.responseProtocol.code) {
                // @todo: fallbacks? multiple valid codes?
                return false;
            }
            // data is valid
            let validData = {};
            if (this.responseProtocol.validator) { // @todo: typechecking ("class that implements....")
                const validator = new this.responseProtocol.validator();
                validData = validator.resolve(response);
                if (validData === false) {
                    return false;
                }
            }
            // data gets parsed
            if (this.responseProtocol.parser && validData) { // @todo: typechecking ("class that implements....")
                const parser = new this.responseProtocol.parser();
                parser.context = this.context;
                validData = parser.parse(validData);
                if (validData === false) {
                    return false;
                }
            }
            this.response.data = validData;
            return true;
        }
        /**
         *
         */
        get payload() {
            return this._payload;
        }
        /**
         *
         */
        get id() {
            return this._id || (new Date().getTime() + Math.random() * 100).toString();
        }
        /**
         *
         */
        get name() {
            return this.constructor.name;
        }
        /**
         *
         */
        get protocolLogic() {
            // TODO: I suspect an error here;
            return this.constructor.protocolLogic || [];
        }
        /**
         *
         */
        get channel() {
            return -1;
        }
        /**
         *
         */
        get layer() {
            return -1;
        }
        /**
         *
         */
        get address() {
            return '';
        }
        get token() {
            return this._token;
        }
        /**
         *
         */
        get status() {
            return this._status;
        }
        /**
         *
         */
        set status(code) {
            if (code !== this._status) {
                this._status = code;
                if (this.onStatusChanged) {
                    this.onStatusChanged(this._status);
                }
            }
        }
        /**
         *
         */
        serialize() {
            return {
                channel: this.channel,
                layer: this.layer,
                payload: this.payload,
                response: this.response,
                status: this.status,
                _commandName: this.constructor['name'],
                _objectParams: this._objectParams,
                _stringParamsArray: this._stringParamsArray
            };
        }
        /**
         *
         */
        populate(cmdVO, id) {
            this._stringParamsArray = cmdVO._stringParamsArray;
            this._objectParams = cmdVO._objectParams;
            this.response = cmdVO.response;
            this._id = id;
        }
        /**
         *
         */
        toString() {
            let message = '';
            switch (this.status) {
                case IAMCPStatus.Invalid:
                    message = 'Invalid command';
                    break;
                case IAMCPStatus.New:
                    message = 'New command';
                    break;
                case IAMCPStatus.Queued:
                    message = 'Queued command';
                    break;
                case IAMCPStatus.Sent:
                    message = 'Sent command';
                    break;
                case IAMCPStatus.Suceeded:
                    message = 'Succeeded command';
                    break;
                case IAMCPStatus.Failed:
                    message = 'Failed command';
                    break;
            }
            return message;
        }
        /**
         *
         */
        validateParam(signature) {
            let result;
            let param;
            // objectParams parsing
            if (this._objectParams.hasOwnProperty(signature.name)) {
                param = this._objectParams[signature.name];
            }
            else {
                // stringParam parsing
                if (this._stringParamsArray.length > 0) {
                    param = this._stringParamsArray;
                }
                else {
                    return false;
                }
            }
            // filter out undefined object params
            if (param === undefined) {
                return false;
            }
            result = signature.validation.resolve(param, (signature.key || signature.name));
            if (result !== false) {
                signature.validation.resolved = true;
                if (typeof result === 'object' && result.hasOwnProperty('raw') && result.hasOwnProperty('payload')) {
                    signature.payload = result.payload;
                    signature.raw = result.raw;
                }
                else {
                    signature.payload = result;
                }
                return true;
            }
            else {
                return false;
            }
        }
        /**
         *
         */
        validateProtocolLogic() {
            if (!this.protocolLogic) {
                return true;
            }
            let result;
            for (let rule of this.protocolLogic) {
                result = rule.resolve(this.paramProtocol);
                this.paramProtocol = result;
            }
            return true;
        }
        /**
         *
         */
        validateChannel() {
            let result;
            let validator = new PositiveNumberValidatorBetween(1, 9999);
            let param;
            if (this._objectParams.hasOwnProperty('channel')) {
                param = Number(this._objectParams['channel']);
            }
            else {
                param = NaN;
            }
            result = validator.resolve(param);
            if (result !== false) {
                return Number(result);
            }
            // @todo: dispatch error
            return NaN;
        }
        /**
         *
         */
        validateLayer(fallback) {
            let result;
            let validator = new PositiveNumberValidatorBetween(0, 9999);
            let param;
            if (this._objectParams.hasOwnProperty('layer')) {
                param = Number(this._objectParams['layer']);
            }
            else {
                param = fallback || NaN;
            }
            result = validator.resolve(param);
            if (result !== false) {
                return Number(result);
            }
            // @todo: dispatch error
            return 0;
        }
    }
    Command.AbstractCommand = AbstractCommand;
    /**
     *
     */
    function isIAMCPCommand(object) {
        // @todo: better inheritance type checking
        for (let prop in AbstractCommand.prototype) {
            if (object[prop] === undefined) {
                return false;
            }
        }
        return true;
    }
    Command.isIAMCPCommand = isIAMCPCommand;
    /**
     *
     */
    class AbstractOrChannelOrLayerCommand extends AbstractCommand {
        /**
         *
         */
        constructor(params, context) {
            super(params, context);
            let channel = this.validateChannel();
            let layer = this.validateLayer();
            if (channel) {
                this._channel = channel;
                if (layer) {
                    this._layer = layer;
                }
            }
        }
        /**
         *
         */
        get channel() {
            return this._channel || -1;
        }
        /**
         *
         */
        get layer() {
            return this._layer || -1;
        }
        /**
         *
         */
        get address() {
            let address = '';
            if (this.channel && (this.channel > -1)) {
                address = this.channel.toString();
            }
            else {
                return address;
            }
            if (this.layer && (this.layer > -1)) {
                address = `${address}-${this.layer}`;
            }
            return address;
        }
    }
    Command.AbstractOrChannelOrLayerCommand = AbstractOrChannelOrLayerCommand;
    /**
     *
     */
    class AbstractChannelCommand extends AbstractCommand {
        /**
         *
         */
        constructor(params, context) {
            super(params, context);
            let channel = this.validateChannel();
            if (channel) {
                this._channel = channel;
                this._layer = -1;
            }
            else {
                throw new Error('Needs channel'); // @todo: dispatch
            }
        }
        /**
         *
         */
        get channel() {
            return this._channel || -1;
        }
        /**
         *
         */
        get layer() {
            return -1;
        }
        /**
         *
         */
        get address() {
            if (this.channel) {
                return this.channel.toString();
            }
            else {
                return '';
                // @todo throw???
            }
        }
    }
    Command.AbstractChannelCommand = AbstractChannelCommand;
    /**
     *
     */
    class AbstractLayerCommand extends AbstractCommand {
        /**
         *
         */
        constructor(params, context) {
            super(params, context);
            let channel = this.validateChannel();
            let layer = this.validateLayer();
            if (channel && layer) {
                this._channel = channel;
                this._layer = layer;
            }
            else {
                throw new Error('Needs both channel and layer'); // @todo: dispatch
            }
        }
        /**
         *
         */
        get channel() {
            return this._channel || -1;
        }
        /**
         *
         */
        get layer() {
            return this._layer || -1;
        }
        /**
         *
         */
        get address() {
            let address;
            if (this.channel && (this.channel > -1)) {
                address = this.channel.toString();
            }
            else {
                return '';
                // @todo throw???
            }
            if (this.layer && (this.layer > -1)) {
                address = `${address}-${this.layer}`;
            }
            else {
                return '';
                // @todo throw???
            }
            return address;
        }
    }
    Command.AbstractLayerCommand = AbstractLayerCommand;
    /**
     *
     */
    class AbstractChannelOrLayerCommand extends AbstractCommand {
        /**
         *
         */
        constructor(params, context) {
            super(params, context);
            let channel = this.validateChannel();
            let layer = this.validateLayer();
            if (channel) {
                this._channel = channel;
                if (layer) {
                    this._layer = layer;
                }
            }
            else {
                throw new Error('Needs at least channel'); // @todo: dispatch
            }
        }
        /**
         *
         */
        get channel() {
            return this._channel || -1;
        }
        /**
         *
         */
        get layer() {
            return this._layer || -1;
        }
        /**
         *
         */
        get address() {
            let address;
            if (this.channel) {
                address = this.channel.toString();
            }
            else {
                return '';
                // @todo throw???
            }
            if (this.layer && (this.layer > -1)) {
                address = `${address}-${this.layer}`;
            }
            return address;
        }
    }
    Command.AbstractChannelOrLayerCommand = AbstractChannelOrLayerCommand;
    /**
     *
     */
    class AbstractLayerWithFallbackCommand extends AbstractCommand {
        /**
         *
         */
        constructor(params, context) {
            super(params, context);
            let channel = this.validateChannel();
            let layer = this.validateLayer(0);
            if (channel) {
                this._channel = channel;
                this._layer = layer;
            }
            else {
                throw new Error('Needs at least channel, layer will default to 0 if not specified'); // @todo: dispatch
            }
        }
        /**
         *
         */
        get channel() {
            return this._channel || -1;
        }
        /**
         *
         */
        get layer() {
            return this._layer || -1;
        }
        /**
         *
         */
        get address() {
            let address;
            if (this.channel) {
                address = this.channel.toString();
            }
            else {
                return '';
                // @todo throw???
            }
            if (this.layer && (this.layer > -1)) {
                address = `${address}-${this.layer}`;
            }
            return address;
        }
    }
    Command.AbstractLayerWithFallbackCommand = AbstractLayerWithFallbackCommand;
    /**
     *
     */
    class AbstractLayerWithCgFallbackCommand extends AbstractCommand {
        /**
         *
         */
        constructor(params, context) {
            super(params, context);
            let channel = this.validateChannel();
            let layer = this.validateLayer(9999);
            if (channel) {
                this._channel = channel;
                this._layer = layer;
            }
            else {
                throw new Error('Needs at least channel, layer will default to 9999 if not specified'); // @todo: dispatch
            }
        }
        /**
         *
         */
        get channel() {
            return this._channel || -1;
        }
        /**
         *
         */
        get layer() {
            return this._layer || -1;
        }
        /**
         *
         */
        get address() {
            let address;
            if (this.channel) {
                address = this.channel.toString();
            }
            else {
                return '';
                // @todo throw???
            }
            if (this.layer && (this.layer > -1)) {
                address = `${address}-${this.layer}`;
            }
            return address;
        }
    }
    Command.AbstractLayerWithCgFallbackCommand = AbstractLayerWithCgFallbackCommand;
})(Command = exports.Command || (exports.Command = {}));
